### ISSUE TITLE
ShaderEdittor Needs An Undo, And A Redo Button

### ISSUE THREAD

**Issue by realrufans** [markusfisch/ShaderEditor#92]
#### Components
 
My suggestion will require two buttons to be added at the top of the Editor's interface.

#### Proposal

 There are some basic, or necessary features an app needs to have. Although, it doesn't mean every app out there must have every basic feature, but at least have the basic ones that are related, or needed to give better, or good user experience. 

A few days ago, I had a lot of fun using ShaderEdittor. While using the app as an editor for my shaders code, there are some options I  found missing, as these options are;

* Ability to undo
* Ability to redo the undone.

#### Effect of not having those features

Currently, while typing in the editor, and the user mistakenly deletes a line of their vital codes, the action can't be undone because there is no such key to undo the action. It'd be nice to have these two options implemented. 
The two opinions should act similar to the undo, and redo we see on other apps. Here's a brief definition;
* **Undo**: Erases the last change done to the codes. reverting it to an older state.

* **Redo:** The opposite of undo is to redo. The redo command reverses the undo action.

### One more request

 Currently, the top of Editor's interface looks occupied as you can see from the image below.
![Screenshot_20190414-050622.png](https://cdn.steemitimages.com/DQmXFyQgFBnAxRgTpbGf4bnykSqVx6nu3xyF3vMyAvo7BGc/Screenshot_20190414-050622.png)

So, I was wondering if the available icons can be made smaller? So as to have space for Redo, and Undo option. But, if making the icons smaller isn't possible, then I suggest the removal of the **Tab button**.

I understand that the tab button helps the user to modify their codes for easy readability. But, I believe that the space bar on the user's keypad can also be used to create tabs. So, the Redo an undo option is more Important than the Tab option.

#### Mockups / Examples

The first image below shows the Tab button, and how occupied the options are. The second screenshot shows the replacement of the tab button with the suggested options.

![hey1.png](https://cdn.steemitimages.com/DQmavpNguvsp2k8anZhTmMo6MU6bbEyaAPYFGWLRZCJSwhm/hey1.png)

# How can this be done?

I have taken the time to find the sample code. And, I am glad to provide this awesome code by **zeleven**.

```

public class MainActivity extends AppCompatActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Button redoBtn = (Button) findViewById(R.id.redo);
        Button undoBtn = (Button) findViewById(R.id.undo);
        EditText editText = (EditText) findViewById(R.id.edittext);
      
        // pass edittext object to TextViewUndoRedo class
        TextViewUndoRedo helper = new TextViewUndoRedo(edittext);
        
        // call the method from TextViewUndoRedo class
        redoBtn.setOnClickListener(new View.OnClickListener() {
          @Override
          public void onClick(View v) {
            helper.redo(); // perform redo
          }
        });
        undoBtn.setOnClickLisener(new View.OnClickListener() {
          @Override
          public void onClick(View v) {
            helper.undo(); // perform undo
          }
        });
    }
}


```
The complete codes can be found [Here.](https://gist.github.com/zeleven/0cfa738c1e8b65b23ff7df1fc30c9f7e)

# Benefits of Undo & Redo button

I would love to generally state the overview importance of Redo and undo button over the Tab option.

* If the Tab option isn't available, users will be able to create tabs will space keypad.
* If the Redo, and undo option is missing, users who might have mistakenly cleared their codes would have to start typing all over again.

**Comment by markusfisch** [comment]
Wow! That must be the most detailed feature request I've ever got! 😄Thanks a lot for making this! 👍 Very much appreciated!

I didn't know of @zeleven's `TextViewUndoRedo`. Plugged it in and it works quite well!

About the UI, or rather the tab button. Tabs are not spaces and you cannot insert tab characters (ASCII 9) with the space bar. You may insert a lot of spaces (ASCII 32) to get to the depth of indent you want but that isn't a tab character. A tab character is designed to give _one_ level of indent from the left document edge. It's a precise and semantic way of defining an indent. That said, it's of course a matter of personal taste if one likes to use the character designed to define indent or a lot of spaces to build an indent manually. Personally, I think it's much easier to use a tab character in the editor of Shader Editor because editing on a mobile device is hard enough already. It's all too easy to insert the wrong number of spaces for an indent and (to me) it would be just a pain having to build an indent by hammering the space bar. So I don't like to remove that button because (at least) I would miss it very much.

But you're absolutely right with the crowded `Toolbar` and I don't want to add Undo and Redo as menu items because that would make them inaccessible. Instead, I'm playing with the idea of a second tool bar below the `Toolbar` providing the most used tools. It'd cost some screen space what may post a problem when the keyboard is visible. Especially in landscape orientation. So I'm still searching for ideas on how to add those buttons in the best possible way.

**Comment by realrufans** [comment]
You're right about this 😂
``` build an indent by hammering the space bar. ```

**Comment by realrufans** [comment]
Btw, can the feature requested be on a droo down? 
![Screenshot_20190416-174017.png](https://cdn.steemitimages.com/DQmeH67vBGUeFs6cd1LEfArQ4o9EoiZ7YJTrwaWwB561hDs/Screenshot_20190416-174017.png)

Maybe something similar to that ?

**Comment by markusfisch** [comment]
No, a drop down would mean two taps. That's too much. Undo/Redo should always be available with just a tap.

**Comment by fekga** [comment]
I tried out different editors for shaders and for other code related stuff and what they have done is that the keyboard was modified so it included some special keys (and also a lot of non usable characters were removed). Maybe this will be a good option here too.

**Comment by realrufans** [comment]
Hey @fekga hmm... What you said sounds familiar.
I have a browser that automatically adds some function buttons.
Maybe something like this 
![Screenshot_20190417-192935.png](https://cdn.steemitimages.com/DQmWcuRwHktWhqXx9MDjgy81ZnsZ3ZJU2JCsjr1JuveqGQA/Screenshot_20190417-192935.png)

**Comment by ssartell** [comment]
Any update on this feature request?

**Comment by markusfisch** [comment]
Well, I tried the idea of an additional bar at the bottom of the screen in a local branch but somehow it didn't work too well. It reduced the code area too much and really didn't look very good. At least in that first, very basic try. So I need to put more work into that.

But I have to say that I really like your suggestion to switch buttons according to context! 👍I will definitely try this!

**Comment by markusfisch** [comment]
Implemented in https://github.com/markusfisch/ShaderEditor/commit/cd69a0db6db6a2f9d23cf2e8dfd56532316c40c7

Available from Version 2.17.0 and should become available anytime soon (it's already available in the Beta channel: https://play.google.com/apps/testing/de.markusfisch.android.shadereditor).

At the end I only added two more menu entries. I've tried hiding some of the buttons as @ssartell was suggesting but it didn't work too well either. For example, there's a setting that allows a user to run the shader not in the background but in another activity/task. In that case, I can't hide the resolution switcher because there's no hide code button 😬 

But this is probably not the end of the story 😉 I still like the idea of a bottom bar containing all editor related actions ...
